File: desktopclient/src/main/java/com/goexp/galgame/gui/view/search/frombrand/brand/task/BrandSearchTask.java
package com.goexp.galgame.gui.view.search.frombrand.brand.task;

import com.goexp.galgame.common.model.BrandType;
import com.goexp.galgame.gui.db.mongo.query.BrandQuery;
import com.goexp.galgame.gui.model.Brand;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.concurrent.Task;
import javafx.scene.control.TreeItem;

import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static com.mongodb.client.model.Filters.eq;
import static com.mongodb.client.model.Filters.regex;

public class BrandSearchTask {
    public static class ByName extends Task<ObservableList<TreeItem<Brand>>> {

        private String name;

        public ByName(String name) {
            this.name = name;
        }

        @Override
        protected ObservableList<TreeItem<Brand>> call() {


            var list = BrandQuery.tlp.query()
                    .where(regex("name", "^" + name))
                    .list();

            return FXCollections.observableArrayList(makeTree(list));
        }

        private List<TreeItem<Brand>> makeTree(List<Brand> newValue) {
            return newValue.stream()
                    .collect(Collectors.groupingBy(b -> b.comp != null && b.comp.length() > 0 ? b.comp : ""))
                    .entrySet().stream()
                    .sorted(Comparator.comparing(Map.Entry::getKey))
                    .map(entry -> {
                        var comp = new Brand();
                        comp.comp = entry.getKey();


                        var rootItem = new TreeItem<Brand>();
                        rootItem.setValue(comp);
                        rootItem.setExpanded(true);


                        var brandNodes = entry.getValue().stream()
                                .map(brand -> {
                                    var item = new TreeItem<Brand>();
                                    item.setValue(brand);
                                    return item;
                                }).collect(Collectors.toUnmodifiableList());


                        rootItem.getChildren().setAll(brandNodes);

                        return rootItem;
                    }).collect(Collectors.toUnmodifiableList());
        }
    }

    public static class ByType extends Task<ObservableList<TreeItem<Brand>>> {

        private BrandType type;

        public ByType(BrandType type) {
            this.type = type;
        }

        @Override
        protected ObservableList<TreeItem<Brand>> call() {


            List<Brand> list;
            if (type == BrandType.ALL) {
                list = BrandQuery.tlp.query().list();
            } else {
                list = BrandQuery.tlp.query()
                        .where(eq("type", type.getValue()))
                        .list();

            }

            return FXCollections.observableArrayList(makeTree(list));
        }

        private List<TreeItem<Brand>> makeTree(List<Brand> newValue) {
            return newValue.stream()
                    .collect(Collectors.groupingBy(b -> b.comp != null && b.comp.length() > 0 ? b.comp : ""))
                    .entrySet().stream()
                    .sorted(Comparator.comparing(Map.Entry::getKey))
                    .map(entry -> {
                        var comp = new Brand();
                        comp.comp = entry.getKey();


                        var rootItem = new TreeItem<Brand>();
                        rootItem.setValue(comp);
                        rootItem.setExpanded(true);


                        var brandNodes = entry.getValue().stream()
                                .map(brand -> {
                                    var item = new TreeItem<Brand>();
                                    item.setValue(brand);
                                    return item;
                                }).collect(Collectors.toUnmodifiableList());


                        rootItem.getChildren().setAll(brandNodes);

                        return rootItem;
                    }).collect(Collectors.toUnmodifiableList());
        }
    }

    public static class ByComp extends Task<ObservableList<TreeItem<Brand>>> {

        private String name;

        public ByComp(String name) {
            this.name = name;
        }

        @Override
        protected ObservableList<TreeItem<Brand>> call() {


            var list = BrandQuery.tlp.query()
                    .where(regex("comp", name))
                    .list();

            return FXCollections.observableArrayList(makeTree(list));
        }

        private List<TreeItem<Brand>> makeTree(List<Brand> newValue) {
            return newValue.stream()
                    .collect(Collectors.groupingBy(b -> b.comp != null && b.comp.length() > 0 ? b.comp : ""))
                    .entrySet().stream()
                    .sorted(Comparator.comparing(Map.Entry::getKey))
                    .map(entry -> {
                        var comp = new Brand();
                        comp.comp = entry.getKey();


                        var rootItem = new TreeItem<Brand>();
                        rootItem.setValue(comp);
                        rootItem.setExpanded(true);


                        var brandNodes = entry.getValue().stream()
                                .map(brand -> {
                                    var item = new TreeItem<Brand>();
                                    item.setValue(brand);
                                    return item;
                                }).collect(Collectors.toUnmodifiableList());


                        rootItem.getChildren().setAll(brandNodes);

                        return rootItem;
                    }).collect(Collectors.toUnmodifiableList());
        }
    }


}

File: desktopclient/src/main/scala/com/goexp/galgame/gui/task/brand/BrandSearchTask.scala
package com.goexp.galgame.gui.task.brand

import java.util

import com.goexp.common.util.Strings
import com.goexp.galgame.common.model.BrandType
import com.goexp.galgame.gui.db.mongo.Query.BrandQuery
import com.goexp.galgame.gui.model.Brand
import com.mongodb.client.model.Filters
import com.mongodb.client.model.Filters.regex
import javafx.concurrent.Task
import javafx.scene.control.TreeItem

import scala.collection.JavaConverters._

object BrandSearchTask {
  private def makeTree(brands: util.List[Brand]) =

    brands.asScala.toStream
      .groupBy(b => if (Strings.isNotEmpty(b.comp)) b.comp else "").toStream
      .sortBy({ case (key, _) => key })

      .map({ case (k, v) =>
        val comp = new Brand
        comp.comp = k
        val rootItem = new TreeItem[Brand]
        rootItem.setValue(comp)
        rootItem.setExpanded(true)

        val brandNodes = v.map(brand => {
          val item = new TreeItem[Brand]
          item.setValue(brand)
          item
        }).asJava

        rootItem.getChildren.setAll(brandNodes)
        rootItem
      })
      .asJava

  class ByName(private[this] val name: String) extends Task[util.List[TreeItem[Brand]]] {
    override protected def call = {
      val list = BrandQuery.tlp.query.where(regex("name", "^" + name)).list
      makeTree(list)
    }
  }

  class ByType(private[this] val `type`: BrandType) extends Task[util.List[TreeItem[Brand]]] {
    override protected def call = {
      val list = if (`type` eq BrandType.ALL)
        BrandQuery.tlp.query.list
      else
        BrandQuery.tlp.query.where(Filters.eq("type", `type`.value)).list

      makeTree(list)
    }
  }

  class ByComp(private[this] val name: String) extends Task[util.List[TreeItem[Brand]]] {
    override protected def call = {
      val list = BrandQuery.tlp.query.where(regex("comp", name)).list
      makeTree(list)
    }
  }

}